microservices/object-identifer-strategy.md

```diff
@@ -0,0 +1,140 @@
+
+## URNs
+The use of [Uniform Resource Names](https://en.wikipedia.org/wiki/Uniform_Resource_Name) (URNs) was considered, but rejected. URNs allow the identity of a RESTful resource (object) to be separated from its location. A good analogy is to think of how we generally refer to people by their name rather than their home address, because the latter can change. Using URNs requires the use of a URN resolver, which is like an address book that looks up the URN it is given and produces the [Uniform Resource Identifier](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier) (URI) where that resource can be found. The use of a URN resolver increases network traffic because every reference to a resource requires an additional call to look up its location.
+
+There is currently no clear requirement for the SDC platform that justifies the additional development and operational complexity that would be required to fully implement URNs. If a requirement emerges later it would be easy to add to the strategy proposed here by prefixing all existing UUIDs with a URN namespace identifier (NID) such as `uk.gov.ons.<type>`
+
+## Worked Example
+The example below illustrate how the use of UUIDs to identify objects would work in practice. Rather than get bogged down in the detail the important thing is to note how the UUIDs are passed around between the three collaborating microservices.
+
+For this example the Collection Instrument service needs to load an offline collection instrument (spreadsheet) into the system so it's associated with the correct reporting unit. The RAS back office UI makes a request to the Survey service to get the list of known surveys so it can present their names to the user:
+
+`GET https://surveysvc/surveys`
+
+```json 
+[{
+    "id": "cb0711c3-0ac8-41d3-ae0e-567e5ea1ef87",
+    "shortName": "BRES"
+}]
+```
+
+The user selects the survey applicable to the collection instrument they want to load e.g. BRES. At this point the application has the unique identifier for the BRES survey object and can ask the Collection Exercise service for the list of collection exercises for that survey:
+
+`GET https://collectionexercisesvc/collectionexercises/survey/cb0711c3-0ac8-41d3-ae0e-567e5ea1ef87`
+
+```json
+[{
+    "id": "2c82b85a-2a03-434d-8019-eae05b01c6f4",
+    "name": "201601",
+    "scheduledExecutionDateTime": "2017-05-10 11:55:00Z"
+},
+{
+    "id": "...",
+    "name": "...",
+    "scheduledExecutionDateTime": "..."
+}]
+```
+
+The collection exercise details are displayed to the user, who selects the relevant collection exercise. The Collection Instrument service can now persist the spreadsheet by making an HTTP PUT request to update the relevant collection instrument resource. Note how the request payload contains the UUID for the collection exercise retrieved in the previous step:
+
+`PUT https://collectioninstrumentsvc/collectioninstruments/6A33A6C9-F01F-4224-81D6-362C211DFD31`
```

Comment:
Where does the `6A33A6C9-F01F-4224-81D6-362C211DFD31` identifier come from?

Reply:
It's not explicit in the example for the sake of brevity, but the unwritten assumption is that it's previously been retrieved from the Collection Instrument service via a GET request.